### PR TITLE
add additional eslint rules from the base ruleset

### PR DESCRIPTION
**Pull request by uidoyen**
Closes: [RHOAIENG-1735](https://issues.redhat.com/browse/RHOAIENG-1735)
## Description

This PR intends to enforce the following rules with an error severity:
```
"no-else-return": ["error", { "allowElseIf": false }],
"no-param-reassign": [
      "error",
      { "props": true, "ignorePropertyModificationsFor": ["acc", "e"] }
    ],
"react/prop-types": "off",
    "array-callback-return": ["error", { "allowImplicit": true }],
    "prefer-template": "error",
    "no-lone-blocks": "error",
    "no-lonely-if": "error",
    "no-promise-executor-return": "error",
    "no-restricted-globals": [
      "error",
      {
        "name": "isFinite",
        "message": "Use Number.isFinite instead https://github.com/airbnb/javascript#standard-library--isfinite"
      },
      {
        "name": "isNaN",
        "message": "Use Number.isNaN instead https://github.com/airbnb/javascript#standard-library--isnan"
      }
    ],
    "no-sequences": "error",
    "no-undef-init": "error",
    "no-unneeded-ternary": ["error", { "defaultAssignment": false }],
    "no-useless-computed-key": "error",
    "no-useless-return": "error",
    "symbol-description": "error",
    "yoda": "error",
    "func-names": "warn"
```
## How Has This Been Tested?

`npm run test`

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/api/k8s/servingRuntimes.ts [hunk 1]
```diff
@@ -114,25 +114,25 @@ export const assembleServingRuntime = (
...
+      volumeMounts,
+    };
+  });
 
-  if (isModelMesh) {
-    servingRuntime.spec.tolerations = tolerations;
-  }
+  updatedServingRuntime.spec.tolerations = isModelMesh
+    ? tolerations
+    : servingRuntime.spec.tolerations;
```
Comment (christianvogt) [inline frontend/src/api/k8s/servingRuntimes.ts:135]: Why this change?
Wouldn't it just be:
```
  if (isModelMesh) {
    updatedServingRuntime.spec.tolerations = tolerations;
  }
```

File: frontend/src/concepts/areas/utils.ts [hunk 2]
```diff
@@ -14,7 +14,10 @@ const getFlags = (dashboardConfigSpec: DashboardConfigKind['spec']): FlagState =
...
       const value = flags[key as FeatureFlag];
       if (isFeatureFlag(key, value)) {
-        flagState[key] = key.startsWith('disable') ? !value : value;
+        return {
+          ...flagState,
+          [key]: key.startsWith('disable') ? !value : value,
+        };
       }
       return flagState;
     }, {}),
```
Comment (christianvogt) [inline frontend/src/concepts/areas/utils.ts:23]: This is one of those cases you can use the `acc` variable since it is an ignored property in the rule.

File: frontend/src/pages/exploreApplication/EnableModal.tsx [hunk 3]
```diff
@@ -54,15 +54,16 @@ const EnableModal: React.FC<EnableModalProps> = ({ selectedApp, shown, onClose }
   React.useEffect(() => {
     if (validationInProgress && validationStatus === EnableApplicationStatus.SUCCESS) {
       setValidationInProgress(false);
-      selectedApp.spec.isEnabled = true;
-      selectedApp.spec.shownOnEnabledPage = true;
+      const updatedSelectedApp = { ...selectedApp };
+      updatedSelectedApp.spec.isEnabled = true;
+      updatedSelectedApp.spec.shownOnEnabledPage = true;
```
Comment (christianvogt) [inline frontend/src/pages/exploreApplication/EnableModal.tsx:59]: This changes the current logic.
The modification in place is expected.

However this approach is completely wrong. We should not be modifying the object in place but rather notifying the called so whomever owns the object can modify it and / or react to the situation.

The simplest solution today would be to disable the rule here and add a `TODO` comment.

File: frontend/src/pages/notebookController/screens/server/EnvironmentVariablesField.tsx [hunk 4]
```diff
@@ -39,8 +39,9 @@ const EnvironmentVariablesField: React.FC<EnvironmentVariablesFieldProps> = ({
   }, [variable.type]);
 
   const handleSecretChange = (checked: boolean) => {
-    variable.type = checked ? 'password' : 'text';
-    setVariableType(variable.type);
+    const updatedVariable = { ...variable };
+    updatedVariable.type = checked ? 'password' : 'text';
+    setVariableType(updatedVariable.type);
```
Comment (christianvogt) [inline frontend/src/pages/notebookController/screens/server/EnvironmentVariablesField.tsx:44]: This is a change in logic. Likely needs to call `onUpdateVariable`.

The simplest solution today would be to disable the rule here and add a TODO comment.

File: frontend/src/utilities/__tests__/useFetchState.spec.ts [hunk 5]
```diff
@@ -108,7 +108,7 @@ describe('useFetchState', () => {
     expect(result[1][1]).toBe(true);
 
     // rerender but with a promise that doens't resolve
-    renderResult.rerender(() => new Promise(() => null), [11], {
+    renderResult.rerender(() => Promise.resolve([] as number[]), [11], {
```
Comment (christianvogt) [inline frontend/src/utilities/__tests__/useFetchState.spec.ts:111]: The comment clearly states `rerender but with a promise that doens't resolve`. This change now renders with a resolved promise. Please revert and address any linting issues another way to keep the integrity of the test case.

File: frontend/src/__tests__/cypress/cypress/utils/failEarly.ts [hunk 6]
```diff
@@ -34,7 +34,7 @@ export const failEarly = (): void => {
     }
     cy.wrap(failedTest[specName]).should('be.undefined');
   });
-  afterEach(function () {
+  afterEach(function handleFailedTest() {
```
[the remark this answers is not in the document]
Comment (uidoyen) [reply frontend/src/__tests__/cypress/cypress/utils/failEarly.ts:37]: Arrow function complaining about `this` context here 
Screenshot 2024-01-30 at 3 41 25 AM

File: frontend/src/api/k8s/servingRuntimes.ts (hunk 1 above)
Comment (uidoyen) [reply frontend/src/api/k8s/servingRuntimes.ts:135]: @christianvogt Yes, this can also be done. I am trying to understand, would the new syntax can cause the issue?

File: frontend/src/utilities/__tests__/useFetchState.spec.ts (hunk 5 above)
Comment (uidoyen) [reply frontend/src/utilities/__tests__/useFetchState.spec.ts]: @christianvogt Tried fixing with the following solution. Please let me know if there is a better solution.

```
    renderResult.rerender(
      () =>
        new Promise<never>(() => {
          /* no action */
        }),
      [11],
      {
        initialPromisePurity: true,
      },
    );
  ```

File: frontend/src/concepts/areas/utils.ts (hunk 2 above)
Comment (uidoyen) [reply frontend/src/concepts/areas/utils.ts]: Thanks for suggestion!

File: frontend/src/api/k8s/servingRuntimes.ts (hunk 1 above)
Comment (christianvogt) [reply frontend/src/api/k8s/servingRuntimes.ts]: Looking at the old code, it didn't do an assignment if `!isModelMesh` so I'm trying to keep the logic as close to the same when making these lint changes.
Is the else case you added not simply assigning the exact same value that is already present since `updatedServingRuntime` starts off as `servingRuntime` ? If so, this assignment is unnecessary.

File: frontend/src/utilities/__tests__/useFetchState.spec.ts (hunk 5 above)
Comment (christianvogt) [reply frontend/src/utilities/__tests__/useFetchState.spec.ts]: Yes the above seems fine as the logic is the same. The lint rule of no empty blocks kicks in where you need a comment in the block.

File: frontend/src/__tests__/cypress/cypress/utils/failEarly.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/src/__tests__/cypress/cypress/utils/failEarly.ts:37]: Oh I didn't realize this when I made the comment. The original change is fine to keep.

File: frontend/src/pages/exploreApplication/EnableModal.tsx [hunk 7]
```diff
@@ -54,15 +54,19 @@ const EnableModal: React.FC<EnableModalProps> = ({ selectedApp, shown, onClose }
...
+      /* eslint-enable no-param-reassign */
+
       onClose();
     }
     if (validationInProgress && validationStatus === EnableApplicationStatus.FAILED) {
       setValidationInProgress(false);
       setPostError(validationErrorMessage);
     }
-  }, [onClose, selectedApp.spec, validationErrorMessage, validationInProgress, validationStatus]);
+  }, [onClose, selectedApp, validationErrorMessage, validationInProgress, validationStatus]);
```
Comment (christianvogt) [inline frontend/src/pages/exploreApplication/EnableModal.tsx:69]: Undo this change. While the logic in this file is fishy, we don't need to add unnecessary changes to this PR.

File: frontend/src/utilities/imageUtils.ts [hunk 8]
```diff
@@ -61,10 +63,11 @@ export const getVersion = (version?: string, prefix?: string): string => {
   if (!version) {
     return '';
   }
+
   const versionString =
-    version.startsWith('v') || version.startsWith('V') ? version.slice(1) : version;
+    version && (version.startsWith('v') || version.startsWith('V')) ? version.slice(1) : version;
```
Comment (christianvogt) [inline frontend/src/utilities/imageUtils.ts:68]: Unnecessary because on line 63 `!version` is already checked. Undo this change.

File: frontend/src/utilities/imageUtils.ts [hunk 9]
```diff
@@ -61,10 +63,11 @@ export const getVersion = (version?: string, prefix?: string): string => {
...
 
-  return `${prefix ? prefix : ''}${versionString}`;
+  return (prefix || '') + (versionString || '');
```
Comment (christianvogt) [inline frontend/src/utilities/imageUtils.ts:70]: Prefer templates over concatenation:
```suggestion
  return `${prefix || ''}${versionString}`;
```

File: frontend/src/utilities/useWatchNotebooksForUsers.tsx [hunk 10]
```diff
@@ -54,10 +54,13 @@ const useWatchNotebooksForUsers = (
...
-            notebookMap[value.name] = value.data;
-            return notebookMap;
-          }, {} as UsernameMap<NotebookRunningState>);
+          const newNotebooks = successes.reduce(
+            (notebookMap, { value }) => ({
+              ...notebookMap,
+              [value.name]: value.data,
+            }),
+            {} as UsernameMap<NotebookRunningState>,
+          );
```
Comment (christianvogt) [inline frontend/src/utilities/useWatchNotebooksForUsers.tsx:63]: Use variable name `acc`.
```suggestion
          const newNotebooks = successes.reduce((acc, { value }) => {
            acc[value.name] = value.data;
            return acc;
          }, {} as UsernameMap<NotebookRunningState>);
```

File: frontend/src/utilities/utils.ts [hunk 11]
```diff
@@ -68,7 +68,7 @@ export const getLabelColorForDocType = (
       return 'grey';
   }
 };
-export const combineCategoryAnnotations = (doc: OdhDocument, app: OdhApplication): void => {
+export const combineCategoryAnnotations = (doc: OdhDocument, app: OdhApplication): OdhDocument => {
```
Comment (christianvogt) [inline frontend/src/utilities/utils.ts:71]: You cannot simply change the behavior of a function like this without addressing the caller.
I think there's a bug in the code calling this function.

Best to add a new value to the rule option support ignoring the rule.
`"ignorePropertyModificationsForRegex": ["^assignable[A-Z]"]`

Then rename `doc` to `assignableDoc`.

File: frontend/src/utilities/useDraggableTable.ts [hunk 12]
```diff
@@ -29,12 +29,13 @@ const useDraggableTable = (
...
-    const draggedItemId = evt.currentTarget.id;
+    const updatedEvt = { ...evt };
+    updatedEvt.dataTransfer.effectAllowed = 'move';
+    updatedEvt.dataTransfer.setData('text/plain', updatedEvt.currentTarget.id);
+    const draggedItemId = updatedEvt.currentTarget.id;
 
-    evt.currentTarget.classList.add(styles.modifiers.ghostRow);
-    evt.currentTarget.setAttribute('aria-pressed', 'true');
+    updatedEvt.currentTarget.classList.add(styles.modifiers.ghostRow);
+    updatedEvt.currentTarget.setAttribute('aria-pressed', 'true');
```
Comment (christianvogt) [inline frontend/src/utilities/useDraggableTable.ts:38]: Drag events work this way and must be modified.

Best to add a new value to the rule option support ignoring the rule.
"ignorePropertyModificationsForRegex": ["^assignable[A-Z]"]

Then rename `evt` to `assignableEvent`.
